File: python/testsum/testsum.py
import sys, getopt, os, time
import binascii
import zlib
import hashlib
from pathlib import Path
if sys.version_info < (3, 6):
    import sha3

def printf(format, *args):
    sys.stdout.write(format % args)

def list_files_recursive(path,hashsize):
    """
    Function that receives as a parameter a directory path
    :return list_: File List and Its Absolute Paths
    """

    import os

    #files = []

    # r = root, d = directories, f = files
    for r, d, f in os.walk(path):
        for dir in d:
            newdir = path+"\\"+dir
            if os.path.isdir(newdir):
                #print (newdir)
                list_files_recursive(newdir,hashsize)
        for file in f:
            newfilename = path+"\\"+file
            newfile = Path(newfilename)
            if newfile.exists():
                stinfo = os.stat(newfilename)
                file_size = os.path.getsize(newfilename);
                if file_size < 1024*1024*512:
                    with open(newfilename, 'rb') as f:
                        bufdata = f.read()
                        f.close()
                    crc32 = hex(binascii.crc32(bufdata))
                    md5s = hashlib.md5(bufdata).hexdigest()
                    hash_object = hashlib.sha1(bufdata)
                    hex_dig = hash_object.hexdigest()
                    sha2s = ''
                    sha3s = ''
                    if hashsize == 224:
                         hash_object = hashlib.sha224(bufdata)
                         sha2s = hash_object.hexdigest()
                         hash_object = hashlib.sha3_224(bufdata)
                         sha3s = hash_object.hexdigest()
                    elif hashsize == 256:
                         hash_object = hashlib.sha256(bufdata)
                         sha2s = hash_object.hexdigest()
                         hash_object = hashlib.sha3_256(bufdata)
                         sha3s = hash_object.hexdigest()
                    elif hashsize == 384:
                         hash_object = hashlib.sha384(bufdata)
                         sha2s = hash_object.hexdigest()
                         hash_object = hashlib.sha3_384(bufdata)
                         sha3s = hash_object.hexdigest()
                    elif hashsize == 512:
                         hash_object = hashlib.sha512(bufdata)
                         sha2s = hash_object.hexdigest()
                         hash_object = hashlib.sha3_512(bufdata)
                         sha3s = hash_object.hexdigest()
                    print(f"filename=\"{newfilename}\"  crc32=\"{crc32}\"  md5=\"0x{md5s}\"  sha1=\"0x{hex_dig}\"  sha2({hashsize})=\"0x{sha2s}\"  sha3({hashsize})=\"0x{sha3s}\"")
                    #files.append(os.path.join(r, file))

    #lst = [file for file in files]
    #return lst
    return

def main(argv):
   inputdir = ''
   hashsize = 224
   try:
      opts, args = getopt.getopt(argv,"hi:s:",["idir=","size="])
   except getopt.GetoptError:
      print ('testsum.py -i <inputdir> -s <hashsize>')
      sys.exit(2)
   for opt, arg in opts:
      if opt == '-h':
         print ('Build checksums per any file, version 1.0')
         print ('Copyright (C) 2021 Dmitry Stefankov. All Rights Reserved.')
         print ('Usage: testsum.py -i <inputdir> -s <hashsize>')
         sys.exit()
      elif opt in ("-i", "--idir"):
         inputdir = arg
      elif opt in ("-s", "--size"):
         hashsize = int(arg)
   print ('Input dir: ', inputdir)
   print ('Hashsize:  ', hashsize)

   list_files_recursive(inputdir,hashsize)

if __name__ == "__main__":
   main(sys.argv[1:])

File: lua/testsum/testsum.lua
-- Build checksums per any file, version 1.0
-- Copyright (C) 2022 Dmitry Stefankov. All Rights Reserved.
-- Usage: testsum.lua -i <inputdir>

if #arg < 2 then
  print "Build checksums per any file, version 1.0"
  print "Copyright (C) 2022 Dmitry Stefankov. All Rights Reserved."
  print "Usage: testsum.lua -i <dirname>"
  return 1
end

require 'lfs'

local nums = {}



-- DIR_SEP="\" --should be "/" for Unix platforms (Linux and Mac)
DIR_SEP="/"

printf = function(s,...)
           return io.write(s:format(...))
         end -- function


local indirname = ""

i = 1
while( i < #arg ) do
  if arg[i] == '-i' then
    i = i + 1
    indirname = arg[i]
    print("dirname: ", indirname)
  end
  i = i + 1
end

function GetFileExtension(path)
    return path:match("^.+(%..+)$")
end

function get_file_name(file)
      return file:match("^.+/(.+)$")
end


function FastBytesToString(bytes)
  s = {}
  for i = 1, getn(bytes) do
    s[i] = strchar(bytes[i])
  end
  return concat(s)
end


function crc32(buf, size)
  local crc = 0xFFFFFFFF
  local table = {}
  local rem, c
 
  -- calculate CRC-table
  for i = 0, 0xFF do
    rem = i
    for j = 1, 8 do
      if (rem & 1 == 1) then
        rem = rem >> 1
        rem = rem ~ 0xEDB88320
      else
        rem = rem >> 1
      end
    end
    table[i] = rem
  end
 
   for x = 1, size do
    c = string.byte(buf,x)
    crc = (crc >> 8) ~ table[(crc & 0xFF) ~ c]
  end
  return crc ~ 0xFFFFFFFF
end

function GetFileSize( filename )
    local fp = io.open( filename, "rb" )
    if fp == nil then 
 	return nil 
    end
    local filesize = fp:seek( "end" )
    fp:close()
    return filesize
end


function browseFolder(root)
	for entity in lfs.dir(root) do
		if entity~="." and entity~=".." then
			local fullPath=root..DIR_SEP..entity
			--print("root: "..root..", entity: "..entity..", mode: "..(lfs.attributes(fullPath,"mode")or "-")..", full path: "..fullPath)
			local mode=lfs.attributes(fullPath,"mode")
			if mode=="file" then
				--this is where the processing happens. I print the name of the file and its path but it can be any code
				-- print(root.." > "..entity)
				-- print("file:", fullPath )
				filesize = GetFileSize(fullPath)
                                file_attributes = lfs.attributes(fullPath)
                                filename = get_file_name(fullPath)
                                -- printf("filename=\"%s\"  filesize=\"%s\"  datetime=\"%s\"\n",filename,filesize,os.date("%c",file_attributes.modification))
                                -- printf("filename=\"%s\"  filesize=\"%s\"  datetime=\"%s\"\n",filename,filesize,os.date("%a %b %d %X %Y",file_attributes.modification))
                                filelen = tonumber(filesize)
                                if filelen ~= nil and filelen < 4096*1024*0124 then
				   infile = assert(io.open(fullPath,"rb"))
                                   in_bytes = infile:read("*all")
				   infile:close()
                                   crc32_sum = crc32(in_bytes,filelen)
                                   printf("filename=\"%s\" crc32=\"0x%X\"\n",fullPath,crc32_sum)
                                else 
                                  -- printf("filename=\"%s\"  filesize=\"%s\"\n",filename,filesize)
                                end
			elseif mode=="directory" then
			        -- print("directory ->", fullPath )
			        -- print(fullPath )
				browseFolder(fullPath)
			end
		end
	end
end

print("input dirname:",indirname)
-- print(indirname)
browseFolder(indirname)
